### ISSUE TITLE
Clang Static Analyzer incorrectly analyzes overloaded arrow operator

### ISSUE THREAD

**Issue by xbgnx** [llvm/llvm-project#61341]
Here is an example
```
#include <iostream>

template<typename T>
struct let {
  T* ptr;
  let(T* ptr): ptr(ptr) {}
  let(let& other): ptr(other.ptr){
    other.ptr = nullptr;
  }
  ~let(){
    delete ptr;
  }
  T* operator->(){
    return ptr;
  }
};

struct X {
    int field;
};
int main(){
    let obj = new X{.field = 123};
    let obj2 = obj;
    std::cout << obj2->field;
    std::cout << obj->field;
}
```
Clang Static Analyzer doesn't give any warnings (where is obviously a null deference error) but if I change line "obj->field" to "obj.ptr->field" I'll get the expected error  ([godbolt](https://godbolt.org/z/vEs67M544))
```
<source>:26:18: warning: Access to field 'field' results in a dereference of a null pointer (loaded from field 'ptr') [core.NullDereference]
    std::cout << obj.ptr->field;
                 ^~~~~~~~~~~~~~
1 warning generated.
Compiler returned: 0
```
